v2/stream/new/flow.go

```diff
@@ -0,0 +1,193 @@
+package new
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/vmihailenco/msgpack/v5"
+)
+
+var initializeTimeout = 3 * time.Second
+var authRetryDelayMultiplier = time.Second
+var authRetryCount = 15
+
+// initialize performs the initial flow:
+// 1. wait to be welcomed
+// 2. authenticates (and waits for the response)
+// 3. subscribes (and waits for the response)
+//
+// If it runs into retriable issues with authentication it retries for a while
+func (c *client) initialize(ctx context.Context) error {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, initializeTimeout)
+	defer cancel()
+
+	if err := c.readConnected(ctxWithTimeout); err != nil {
+		return err
+	}
+
+	var retryErr error
+	// If we run into a retriable error during auth we should try
+	// again after a delay
+	for i := 0; i < authRetryCount; i++ {
+		if i > 0 {
+			n := i
+			if n > 10 {
+				n = 10
+			}
+			sleepDuration := authRetryDelayMultiplier * time.Duration(n*n)
+			c.logger.Infof("datav2stream: retring auth in %s, attempt %d/%d", sleepDuration, i+1, authRetryCount+1)
+			time.Sleep(sleepDuration)
+		}
+		ctxWithTimeout, cancel := context.WithTimeout(ctx, initializeTimeout)
+		defer cancel()
+		if err := c.writeAuth(ctxWithTimeout); err != nil {
+			return err
+		}
+
+		ctxWithTimeoutResp, cancelResp := context.WithTimeout(ctx, initializeTimeout)
+		defer cancelResp()
+		canBeRetried, err := c.readAuthResponse(ctxWithTimeoutResp)
+		retryErr = err
+		if err == nil {
+			break
+		}
+		if !canBeRetried {
+			return err
+		}
+	}
+
+	if retryErr != nil {
+		return retryErr
+	}
+
+	ctxWithTimeoutWriteSub, cancelWriteSub := context.WithTimeout(ctx, initializeTimeout)
+	defer cancelWriteSub()
+	if err := c.writeSub(ctxWithTimeoutWriteSub); err != nil {
+		return err
+	}
+
+	ctxWithTimeoutReadSub, cancelReadSub := context.WithTimeout(ctx, initializeTimeout)
+	defer cancelReadSub()
+	if err := c.readSubResponse(ctxWithTimeoutReadSub); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var ErrNoConnected = errors.New("did not receive connected message")
+
+func (c *client) readConnected(ctx context.Context) error {
+	b, err := c.conn.readMessage(ctx)
```

Comment:
Is this not called concurrently with another read in `client.connReader`?

https://pkg.go.dev/nhooyr.io/websocket#Conn
> All methods may be called concurrently except for Reader and Read.

Reply:
`flow.go`'s methods are only called by `initialize`.
`initialize` in turn is only called when `connReader` pinger is not running.

Reply:
You are right. My bad again. 🔫 🔫